### PR TITLE
Constraint improvements

### PR DESCRIPTION
**Pull request by tonyhallett**
## Description

Added three constraint helpers - not/and/or.
___
Changed the type constraint to accept all service identifier types.

Changed following BindingWhenSyntax methods

whenInjectedInto
whenAnyAncestorIs
whenNoAncestorIs

to also accept all service identifier types but also to to accept an array as a shortcut.
___

## Related Issue

#599 

## Motivation and Context

Symbols could not be used in the constraint

## How Has This Been Tested?

TDD, npm run test ( Windows )

## Types of changes

- [x] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the changelog.

### CODE REVIEW

File: src/syntax/constraint_helpers.ts [hunk 1]
```diff
@@ -29,24 +29,74 @@ const taggedConstraint = (key: string | number | symbol) => (value: any) => {
...
 
+    return match;
+};
+
+const symbolMatch = (serviceIdentifier: symbol, matchSymbol: symbol): boolean => {
+    if (serviceIdentifier === matchSymbol) {
+        return serviceIdentifier === matchSymbol;
+    }
+    const matchOnSymbolKey = false;
+    if (matchOnSymbolKey) {
```
Comment (tonyhallett) [inline src/syntax/constraint_helpers.ts:63]: This needs to be determined.  Thoughts please.

**Review by tonyhallett — COMMENTED** [review]
I was also going to change the types for BindingWhenSyntax such that
when / whenAnyAncestorMatches / whenNoAncestorMatches use 
```
    export interface ConstraintFunction extends Function {
        metaData?: Metadata;
        (request: Request | null): boolean;
    }
```

instead of current inline e.g

```
    export interface BindingWhenSyntax<T> {
        when(constraint: (request: Request) => boolean): BindingOnSyntax<T>;
```

Thoughts ?

**Comment by tonyhallett** [comment]
@inversify/maintainers 

Please can this be reviewed.  Only the logic on symbol matching needs to be determined.

**Comment by jkhmtx** [comment]
@tonyhallett Would you like to update the branch with the lastest `master` and reopen this one? I'll be happy to review it.

**Pull request closed without merging**